File: Claude/Intermediárias/gale_shapley/Gale-Shapley.py
def stable_marriage(student_prefs, hospital_prefs):
    n = len(student_prefs)
    free_students = list(range(n))
    hospital_students = [-1] * n
    student_next = [0] * n

    while free_students:
        student = free_students.pop(0)
        student_list = student_prefs[student]
        hospital = student_list[student_next[student]]
        student_next[student] += 1

        if hospital_students[hospital] == -1:
            hospital_students[hospital] = student
        else:
            current_student = hospital_students[hospital]
            hospital_list = hospital_prefs[hospital]
            if hospital_list.index(student) < hospital_list.index(current_student):
                free_students.append(current_student)
                hospital_students[hospital] = student
            else:
                free_students.append(student)

    return [(i, hospital_students[i]) for i in range(n)]

# Example usage
student_prefs = [
    [0, 1, 2],
    [1, 0, 2],
    [0, 1, 2]
]
hospital_prefs = [
    [0, 2, 1],
    [1, 0, 2],
    [0, 1, 2]
]

result = stable_marriage(student_prefs, hospital_prefs)
print("Stable pairings (hospital, student):")
for hospital, student in result:
    print(f"{hospital}: {student}")

File: Claude/Intermediárias/gale_shapley/Gale-Shapley.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

vector<pair<int, int>> stable_marriage(const vector<vector<int>>& student_prefs, const vector<vector<int>>& hospital_prefs) {
    int n = student_prefs.size();
    vector<int> hospital_students(n, -1);
    vector<int> student_next(n, 0);
    queue<int> free_students;

    for (int i = 0; i < n; ++i) {
        free_students.push(i);
    }

    while (!free_students.empty()) {
        int student = free_students.front();
        free_students.pop();

        int hospital = student_prefs[student][student_next[student]];
        student_next[student]++;

        if (hospital_students[hospital] == -1) {
            hospital_students[hospital] = student;
        } else {
            int current_student = hospital_students[hospital];
            auto it1 = find(hospital_prefs[hospital].begin(), hospital_prefs[hospital].end(), student);
            auto it2 = find(hospital_prefs[hospital].begin(), hospital_prefs[hospital].end(), current_student);

            if (it1 < it2) {
                free_students.push(current_student);
                hospital_students[hospital] = student;
            } else {
                free_students.push(student);
            }
        }
    }

    vector<pair<int, int>> result;
    for (int i = 0; i < n; ++i) {
        result.emplace_back(i, hospital_students[i]);
    }
    return result;
}

int main() {
    vector<vector<int>> student_prefs = {
        {0, 1, 2},
        {1, 0, 2},
        {0, 1, 2}
    };
    vector<vector<int>> hospital_prefs = {
        {0, 2, 1},
        {1, 0, 2},
        {0, 1, 2}
    };

    auto result = stable_marriage(student_prefs, hospital_prefs);

    cout << "Stable pairings (hospital, student):" << endl;
    for (const auto& [hospital, student] : result) {
        cout << hospital << ": " << student << endl;
    }

    return 0;
}